farm-bot/codebase/contracts/farm-bot.sol

```diff
@@ -2,62 +2,205 @@
-    uint256 public fpTotalBalance; // total number of "Farm Points" issued by Farm Bot
 
-    address public tokenAddress = 0xf3098223845F37Ffe4B3A066F2D38A0587317269; // mcUSD-Celo
-    StakingRewards public farm = StakingRewards(0x299f31f48D4667a6f68E4331dB05212d57Cc7f80); // alfajores farm for mcUSD-Celo
+    StakingRewards public stakingRewards;
+
+    IERC20 public rewardsToken;
+
+    IUniswapV2Pair public stakingToken; // LP that's being staked
+    IERC20 public stakingToken0; // LP token0
+    IERC20 public stakingToken1; // LP token1
+
+    IUniswapV2Router02 public router; // Router address
+
+    // Paths for swapping; can be updated by owner
+    address[] public path0; // Path to use when swapping rewardsToken to token0. If len < 2, we assume rewardsToken == token0
+    address[] public path1; // Path to use when swapping rewardsToken to token1. If len < 2, we assume rewardsToken == token1
+
+    // Acceptable slippage when swapping/minting LP; can be updated by owner
+    uint256 public slippageNumerator = 99;
+    uint256 public slippageDenominator = 100;
+
+    // Fee awarded to address triggering autocompounding; can be updated by owner
+    uint256 public feeNumerator = 1;
+    uint256 public feeDenominator = 1000; // .1% fee awarded to claimRewards caller
+
+    modifier ensure(uint deadline) {
+        require(deadline >= block.timestamp, 'FarmBot: EXPIRED');
+        _;
+    }
+
+    constructor(
+        address _owner,
+        address _stakingRewards,
+	address _router,
+	address[] memory _path0,
+	address[] memory _path1
+    ) Owned(_owner) {
+        stakingRewards = StakingRewards(_stakingRewards);
+	rewardsToken = stakingRewards.rewardsToken();
+
+	stakingToken = IUniswapV2Pair(address(stakingRewards.stakingToken()));
+	stakingToken0 = IERC20(stakingToken.token0());
+	stakingToken1 = IERC20(stakingToken.token1());
+
+	path0 = _path0;
+	path1 = _path1;
+
+	router = IUniswapV2Router02(_router);
+    }
+
+    function updateFee(uint256 _feeNumerator, uint256 _feeDenominator) external onlyOwner {
+	feeNumerator = _feeNumerator;
+	feeDenominator = _feeDenominator;
+    }
+
+    function updatePaths(address[] calldata _path0, address[] calldata _path1) external onlyOwner {
+	path0 = _path0;
+	path1 = _path1;
+    }
+
+    function updateSlippage(uint256 _slippageNumerator, uint256 _slippageDenominator) external onlyOwner {
+	slippageNumerator = _slippageNumerator;
+	slippageDenominator = _slippageDenominator;
+    }
 
     function getFpAmount(uint256 _lpAmount) public view returns (uint256) {
-        // todo check the rounding on this at some point...
         if (lpTotalBalance == 0) {
             return _lpAmount;
         } else {
-            return _lpAmount * fpTotalBalance / lpTotalBalance;
+            return _lpAmount * totalSupply / lpTotalBalance;
         }
     }
 
+    function getLpAmount(uint256 _fpAmount) public view returns (uint256) {
+	if (totalSupply == 0) {
+	    return 0;
```

Comment:
This is kind of confusing. It may never happen, but as written, if LP and FP balance is 0, `getFpAmount(getLpAmount(x)) !== x` for all  `x > 0`. I'm not sure how we could improve on it, though.

Reply:
Maybe just renaming the two functions to `getFpTradeInValue` and `getLpTradeInValue` would be easier. That way it better describes the purpose of the functions and they don't seem like inverses of one another as much anymore.

Trading in FP for LP when the contract balance is 0 gets you a 1:1 exchange rate, whereas trading in FP for LP when the contract balance is 0 makes no sense and will get you nothing